horovod/spark/common/util.py

```diff
@@ -0,0 +1,530 @@
+        #         mixed SparseVector and DenseVector we categorize it as DenseVector.
+        #
+        #     is_sparse_vector_only:
+        #         If all the rows in the column were sparse vectors.
+        #
+        #     shape:
+        #         Determines the shape of the data in the spark dataframe. It is useful for sparse
+        #         vectors.
+        #
+        #     intermediate_format:
+        #         Specifies if the column need to be converted to a different format so that
+        #         petastorm can read it. It can be one of ARRAY, CUSTOM_SPARSE, or NOCHANGE. It is
+        #         required because petastorm cannot read DenseVector and SparseVectors. We need to
+        #         identify these types and convert them to petastorm compatible type of array.
+
+        metadata[col] = {'spark_data_type': spark_data_type,
+                         'is_sparse_vector_only': is_sparse_vector_only,
+                         'shape': col_shapes[col],
+                         'intermediate_format': convert_to_target,
+                         'max_size': col_max_sizes[col]}
+
+    return metadata
+
+
+def to_petastorm_fn(schema_cols, metadata):
+    ARRAY = constants.ARRAY
+    CUSTOM_SPARSE = constants.CUSTOM_SPARSE
+
+    # Convert Spark Vectors into arrays so Petastorm can read them
+    def to_petastorm(row):
+        import numpy as np
+        from pyspark import Row
+
+        fields = row.asDict().copy()
+        for col in schema_cols:
+            col_data = row[col]
+            intermediate_format = metadata[col]['intermediate_format']
+            if intermediate_format == ARRAY:
+                fields[col] = col_data.toArray().tolist()
+            elif intermediate_format == CUSTOM_SPARSE:
+                # Currently petastorm does not support reading pyspark sparse vector. We put
+                # the indices and values into one array. when consuming the data, we re-create
+                # the vector from this format.
+                size = len(col_data.indices)
+                padding_zeros = 2 * (metadata[col]['max_size'] - len(col_data.indices))
+
+                fields[col] = np.concatenate(
+                    (np.array([size]), col_data.indices, col_data.values,
+                     np.zeros(padding_zeros))).tolist()
+
+        return Row(**fields)
+
+    return to_petastorm
+
+
+def get_simple_meta_from_parquet(store, label_columns, feature_columns, sample_weight_col):
+    train_data_path = store.get_train_data_path()
+    validation_data_path = store.get_val_data_path()
+
+    if not store.exists(train_data_path):
+        raise ValueError("{} path does not exist on the store".format(train_data_path))
+
+    train_data = pq.ParquetDataset(train_data_path, filesystem=store.get_filesystem())
+    schema = train_data.schema.to_arrow_schema()
+
+    train_rows = 0
+    total_byte_size = 0
+    for piece in train_data.pieces:
+        metadata = piece.get_metadata()
```

Comment:
what's the difference between `metadata` and `file_metadata`? It seems equally:
```python3
def get_metadata(self):
        """
        Returns the file's metadata

        Returns
        -------
        metadata : FileMetaData
        """
        f = self.open()
        return f.metadata

    def open(self):
        """
        Returns instance of ParquetFile
        """
        reader = self.open_file_func(self.path)
        if not isinstance(reader, ParquetFile):
            reader = ParquetFile(reader, **self.file_options)
        return reader

```

Reply:
Hmm, after testing seems there's no difference.  Possible there was a difference in older versions of pyarrow, as the API changes a lot, but since we're pinned to newer versions due to Petastorm, I removed the redundant file_metadata.